File: src/utils/features.ts
import type { AppSettings } from '../types';

export type FeatureId =
  | 'analytics'
  | 'bins'
  | 'clipTypes'
  | 'contentClassification'
  | 'notes'
  | 'notifications'
  | 'appLock'
  | 'ocr'
  | 'transcriptions'
  | 'pinning'
  | 'protection'
  | 'queue'
  | 'revisions'
  | 'hud'
  | 'trash'
  | 'transformations'
  | 'activityLog'
  | 'types'
  | 'sources'
  | 'cli'
  | 'help';

export type FeatureSettingKey =
  | 'enableAnalytics'
  | 'enableBins'
  | 'enableClipTypes'
  | 'enableContentClassification'
  | 'enableNotes'
  | 'enableNotifications'
  | 'enableAppLock'
  | 'enableOcr'
  | 'enableTranscriptions'
  | 'enablePinning'
  | 'enableProtection'
  | 'enableQueue'
  | 'enableRevisions'
  | 'enableHud'
  | 'enableTrash'
  | 'enableTransformations'
  | 'enableActivityLog'
  | 'enableTypes'
  | 'enableSources'
  | 'enableCli'
  | 'enableHelp';

export type FeatureGroupId = 'library' | 'discovery' | 'workflow' | 'app';

export interface FeatureGroupDefinition {
  id: FeatureGroupId;
  label: string;
  description: string;
}

export const FEATURE_GROUPS: readonly FeatureGroupDefinition[] = [
  { id: 'library', label: 'Library', description: 'Organize, preserve, and manage clipboard history.' },
  { id: 'discovery', label: 'Intelligence and discovery', description: 'Understand clip contents and browse useful collections.' },
  { id: 'workflow', label: 'Workflow Tools', description: 'Use clips faster and build more capable workflows.' },
  { id: 'app', label: 'App and support', description: 'Control feedback and access supporting tools.' },
] as const;

export interface FeatureDefinition {
  id: FeatureId;
  group: FeatureGroupId;
  settingKey: FeatureSettingKey;
  label: string;
  description: string;
  simple: boolean;
  caution?: string;
}

export const FEATURE_DEFINITIONS: readonly FeatureDefinition[] = [
  { id: 'bins', group: 'library', settingKey: 'enableBins', label: 'Bins', description: 'Organize clips manually or automatically with Smart Bins.', simple: false },
  { id: 'clipTypes', group: 'library', settingKey: 'enableClipTypes', label: 'Clip Types', description: 'Show structural Clip Types and their collections.', simple: false },
  { id: 'contentClassification', group: 'discovery', settingKey: 'enableContentClassification', label: 'Content Classification', description: 'Assign registered Content Types to analyzable text.', simple: true },
  { id: 'notes', group: 'library', settingKey: 'enableNotes', label: 'Notes', description: 'Annotate clips and browse the Noted collection.', simple: false },
  { id: 'notifications', group: 'app', settingKey: 'enableNotifications', label: 'Notifications', description: 'Show interactive capture feedback without interrupting the current workflow.', simple: false },
  {
    id: 'appLock',
    group: 'app',
    settingKey: 'enableAppLock',
    label: 'App Lock',
    description: 'Require authentication before showing clipboard history.',
    simple: false,
    caution: 'The saved passphrase and lock preferences remain available when App Lock is re-enabled.',
  },
  { id: 'ocr', group: 'discovery', settingKey: 'enableOcr', label: 'OCR', description: 'Automatically extract searchable text from copied images.', simple: false },
  { id: 'transcriptions', group: 'discovery', settingKey: 'enableTranscriptions', label: 'Transcriptions', description: 'Create searchable text from copied audio files.', simple: false },
  { id: 'pinning', group: 'library', settingKey: 'enablePinning', label: 'Pinning', description: 'Keep important clips at the top of history.', simple: false },
  {
    id: 'protection',
    group: 'library',
    settingKey: 'enableProtection',
    label: 'Protection',
    description: 'Protect clips from deletion and automatic retention.',
    simple: false,
    caution: 'Previously protected clips remain protected. Re-enable this feature to change them.',
  },
  { id: 'queue', group: 'workflow', settingKey: 'enableQueue', label: 'Copy Queue', description: 'Collect copied text and paste it back in sequence.', simple: false },
  {
    id: 'revisions',
    group: 'library',
    settingKey: 'enableRevisions',
    label: 'Revision History',
    description: 'Keep restorable snapshots before clips change.',
    simple: true,
    caution: 'New edits and Transforms will not be reversible while Revision History is disabled.',
  },
  { id: 'hud', group: 'workflow', settingKey: 'enableHud', label: 'HUD', description: 'Open the compact keyboard-driven clipboard window.', simple: false },
  {
    id: 'trash',
    group: 'library',
    settingKey: 'enableTrash',
    label: 'Trash',
    description: 'Recover deleted clips before they are permanently removed.',
    simple: true,
    caution: 'When disabled, deleting a clip permanently removes it.',
  },
  { id: 'transformations', group: 'workflow', settingKey: 'enableTransformations', label: 'Transformations', description: 'Run text workflows and receive Smart Action suggestions.', simple: false },
  { id: 'analytics', group: 'discovery', settingKey: 'enableAnalytics', label: 'Insights', description: 'Browse active-library composition and recent additions.', simple: false },
  { id: 'activityLog', group: 'app', settingKey: 'enableActivityLog', label: 'Activity', description: 'Record and inspect important app events.', simple: false },
  { id: 'types', group: 'discovery', settingKey: 'enableTypes', label: 'Content Types', description: 'Show recognized Content Types and their collections.', simple: false },
  { id: 'sources', group: 'discovery', settingKey: 'enableSources', label: 'Sources', description: 'Show the applications associated with captured clips and their collections.', simple: false },
  { id: 'cli', group: 'app', settingKey: 'enableCli', label: 'Command-Line Interface', description: 'Use pasted to automate clipboard workflows.', simple: false },
  { id: 'help', group: 'app', settingKey: 'enableHelp', label: 'Help', description: 'Show in-app documentation and its navigation entry.', simple: true },
] as const;

export const FEATURE_SETTING_KEYS = FEATURE_DEFINITIONS.map(({ settingKey }) => settingKey);

export type FeaturePreset = 'full' | 'simple' | 'custom';

export function isFeatureEnabled(settings: AppSettings, featureId: FeatureId): boolean {
  const definition = FEATURE_DEFINITIONS.find(({ id }) => id === featureId);
  return definition ? settings[definition.settingKey] : false;
}

export function featureUpdatesForPreset(preset: Exclude<FeaturePreset, 'custom'>): Partial<AppSettings> {
  return Object.fromEntries(
    FEATURE_DEFINITIONS.map(({ settingKey, simple }) => [settingKey, preset === 'full' || simple]),
  ) as Partial<AppSettings>;
}

export function activeFeaturePreset(settings: AppSettings): FeaturePreset {
  const isFull = FEATURE_DEFINITIONS.every(({ settingKey }) => settings[settingKey]);
  if (isFull) return 'full';
  const isSimple = FEATURE_DEFINITIONS.every(({ settingKey, simple }) => settings[settingKey] === simple);
  return isSimple ? 'simple' : 'custom';
}

export function enabledFeatureRecord(settings: AppSettings): Record<FeatureId, boolean> {
  return Object.fromEntries(
    FEATURE_DEFINITIONS.map(({ id }) => [id, isFeatureEnabled(settings, id)]),
  ) as Record<FeatureId, boolean>;
}

export function featureForRoute(route: string): FeatureId | null {
  if (route.startsWith('clip_type-')) return 'clipTypes';
  if (route.startsWith('type-')) return 'types';
  if (route.startsWith('source-')) return 'sources';
  const tab = route.split(':', 1)[0];
  const routes: Record<string, FeatureId> = {
    sequential: 'queue',
    pinned: 'pinning',
    protected: 'protection',
    notes: 'notes',
    trash: 'trash',
    bin: 'bins',
    analytics: 'analytics',
    transformations: 'transformations',
    activity: 'activityLog',
    help: 'help',
  };
  return routes[tab] ?? null;
}

File: src-tauri/src/features.rs
use crate::db::DbState;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Feature {
    Insights,
    Bins,
    ClipTypes,
    FileFormats,
    ContentClassification,
    Concealment,
    Naming,
    Notes,
    Notifications,
    AppLock,
    Ocr,
    Transcriptions,
    Pinning,
    Protection,
    Queue,
    Revisions,
    Hud,
    Hotkeys,
    Trash,
    Transformations,
    ActivityLog,
    ContentTypes,
    Sources,
    Search,
    Cli,
    Help,
}

impl Feature {
    pub const ALL: [Feature; 26] = [
        Feature::Insights,
        Feature::Bins,
        Feature::ClipTypes,
        Feature::FileFormats,
        Feature::ContentClassification,
        Feature::Concealment,
        Feature::Naming,
        Feature::Notes,
        Feature::Notifications,
        Feature::AppLock,
        Feature::Ocr,
        Feature::Transcriptions,
        Feature::Pinning,
        Feature::Protection,
        Feature::Queue,
        Feature::Revisions,
        Feature::Hud,
        Feature::Hotkeys,
        Feature::Trash,
        Feature::Transformations,
        Feature::ActivityLog,
        Feature::ContentTypes,
        Feature::Sources,
        Feature::Search,
        Feature::Cli,
        Feature::Help,
    ];

    pub const fn setting_key(self) -> &'static str {
        match self {
            Feature::Insights => "enableAnalytics",
            Feature::Bins => "enableBins",
            Feature::ClipTypes => "enableClipTypes",
            Feature::FileFormats => "enableFileFormats",
            Feature::ContentClassification => "enableContentClassification",
            Feature::Concealment => "enableConcealment",
            Feature::Naming => "enableNaming",
            Feature::Notes => "enableNotes",
            Feature::Notifications => "enableNotifications",
            Feature::AppLock => "enableAppLock",
            Feature::Ocr => "enableOcr",
            Feature::Transcriptions => "enableTranscriptions",
            Feature::Pinning => "enablePinning",
            Feature::Protection => "enableProtection",
            Feature::Queue => "enableQueue",
            Feature::Revisions => "enableRevisions",
            Feature::Hud => "enableHud",
            Feature::Hotkeys => "enableHotkeys",
            Feature::Trash => "enableTrash",
            Feature::Transformations => "enableTransformations",
            Feature::ActivityLog => "enableActivityLog",
            Feature::ContentTypes => "enableTypes",
            Feature::Sources => "enableSources",
            Feature::Search => "enableSearch",
            Feature::Cli => "enableCli",
            Feature::Help => "enableHelp",
        }
    }

    pub fn from_setting_key(key: &str) -> Option<Self> {
        Self::ALL
            .into_iter()
            .find(|feature| feature.setting_key() == key)
    }

    pub const fn label(self) -> &'static str {
        match self {
            Feature::Insights => "Insights",
            Feature::Bins => "Bins",
            Feature::ClipTypes => "Clip Types",
            Feature::FileFormats => "File Formats",
            Feature::ContentClassification => "Content Classification",
            Feature::Concealment => "Concealment",
            Feature::Naming => "Naming",
            Feature::Notes => "Notes",
            Feature::Notifications => "Notifications",
            Feature::AppLock => "App Lock",
            Feature::Ocr => "OCR",
            Feature::Transcriptions => "Transcriptions",
            Feature::Pinning => "Pinning",
            Feature::Protection => "Protection",
            Feature::Queue => "Queue",
            Feature::Revisions => "Revision History",
            Feature::Hud => "HUD",
            Feature::Hotkeys => "Hotkeys",
            Feature::Trash => "Trash",
            Feature::Transformations => "Transformations",
            Feature::ActivityLog => "Activity",
            Feature::ContentTypes => "Content Types",
            Feature::Sources => "Sources",
            Feature::Search => "Clip Search",
            Feature::Cli => "CLI",
            Feature::Help => "Help",
        }
    }
}

pub fn setting_value_is_enabled(value: Option<&str>) -> bool {
    !matches!(value.map(str::trim), Some("false") | Some("0"))
}

pub fn is_enabled(db: &DbState, feature: Feature) -> bool {
    let value = db.get_setting(feature.setting_key()).ok().flatten();
    setting_value_is_enabled(value.as_deref())
}

pub fn require(db: &DbState, feature: Feature) -> Result<(), String> {
    if is_enabled(db, feature) {
        Ok(())
    } else {
        Err(format!(
            "{} is disabled in Settings → Functionality",
            feature.label()
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;
    use std::time::{SystemTime, UNIX_EPOCH};

    #[test]
    fn missing_and_legacy_values_remain_enabled() {
        assert!(setting_value_is_enabled(None));
        assert!(setting_value_is_enabled(Some("true")));
        assert!(setting_value_is_enabled(Some("unexpected")));
    }

    #[test]
    fn explicit_false_values_disable_a_feature() {
        assert!(!setting_value_is_enabled(Some("false")));
        assert!(!setting_value_is_enabled(Some(" false ")));
        assert!(!setting_value_is_enabled(Some("0")));
    }

    #[test]
    fn frontend_and_native_setting_keys_are_stable() {
        assert_eq!(Feature::ALL.len(), 26);
        for feature in Feature::ALL {
            assert_eq!(
                Feature::from_setting_key(feature.setting_key()),
                Some(feature)
            );
        }
        assert_eq!(Feature::from_setting_key("unrelatedSetting"), None);
    }

    #[test]
    fn persisted_policy_defaults_on_and_applies_atomic_presets() {
        let nonce = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let path = std::env::temp_dir().join(format!("pasted_feature_policy_{nonce}.db"));
        let db = DbState::new(path.clone()).unwrap();

        assert!(is_enabled(&db, Feature::Bins));
        assert!(is_enabled(&db, Feature::Hotkeys));
        let values = HashMap::from([
            (Feature::Bins.setting_key().to_string(), "false".to_string()),
            (
                Feature::Notes.setting_key().to_string(),
                "false".to_string(),
            ),
            (
                Feature::Hotkeys.setting_key().to_string(),
                "false".to_string(),
            ),
        ]);
        db.save_settings(&values).unwrap();
        assert!(!is_enabled(&db, Feature::Bins));
        assert!(!is_enabled(&db, Feature::Notes));
        assert!(!is_enabled(&db, Feature::Hotkeys));
        assert!(is_enabled(&db, Feature::Pinning));

        drop(db);
        let _ = std::fs::remove_file(path);
    }
}
